src/iosrv/sched.c

```diff
@@ -0,0 +1,542 @@
+		return -DER_INVAL;
+	}
+
+	if (pool_idx >= DSS_POOL_CNT) {
+		D_ERROR("Invalid pool idx: %d\n", pool_idx);
+		return -DER_INVAL;
+	}
+
+	if (pool_idx == DSS_POOL_NET_POLL || pool_idx == DSS_POOL_NVME_POLL) {
+		D_ERROR("Can't throttle network or NVMe poll\n");
+		return -DER_INVAL;
+	}
+
+	sched_throttle[pool_idx] = percent;
+	return 0;
+}
+
+/* #define SCHED_DEBUG */
+static void
+sched_dump_data(struct sched_data *data)
+{
+#ifdef SCHED_DEBUG
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+
+	D_PRINT("XS(%d): comm:%d main:%d. age_net:%u, [%u, %u], "
+		"age_nvme:%u, [%u, %u] new_cycle:%d cycle_started:%d "
+		"total_ults:%u [%u, %u, %u, %u]\n", dx->dx_xs_id,
+		dx->dx_comm, dx->dx_main_xs, cycle->sc_age_net,
+		cycle->sc_age_net_bound[0], cycle->sc_age_net_bound[1],
+		cycle->sc_age_nvme, cycle->sc_age_nvme_bound[0],
+		cycle->sc_age_nvme_bound[1], cycle->sc_new_cycle,
+		cycle->sc_cycle_started, cycle->sc_ults_tot,
+		cycle->sc_ults_cnt[DSS_POOL_IO],
+		cycle->sc_ults_cnt[DSS_POOL_REBUILD],
+		cycle->sc_ults_cnt[DSS_POOL_AGGREGATE],
+		cycle->sc_ults_cnt[DSS_POOL_GC]);
+#endif
+}
+
+#define SCHED_AGE_NET_MIN		32
+#define SCHED_AGE_NET_MAX		512
+#define SCHED_AGE_NVME_MIN		32
+#define SCHED_AGE_NVME_MAX		512
+
+static int
+sched_init(ABT_sched sched, ABT_sched_config config)
+{
+	struct sched_data	*data;
+	struct sched_cycle	*cycle;
+	int			 ret;
+
+	D_ALLOC_PTR(data);
+	if (data == NULL)
+		return ABT_ERR_MEM;
+
+	ret = ABT_sched_config_read(config, 2, &data->sd_event_freq,
+				    &data->sd_dx);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("Failed to read ABT sched config: %d\n", ret);
+		return ret;
+	}
+
+	cycle = &data->sd_cycle;
+	cycle->sc_age_net_bound[0] = SCHED_AGE_NET_MIN;
+	cycle->sc_age_net_bound[1] = SCHED_AGE_NET_MAX;
+	cycle->sc_age_nvme_bound[0] = SCHED_AGE_NVME_MIN;
+	cycle->sc_age_nvme_bound[1] = SCHED_AGE_NVME_MAX;
+
+	ret = ABT_sched_set_data(sched, (void *)data);
+	return ret;
+}
+
+static bool
+need_net_poll(struct sched_cycle *cycle)
+{
+	/* Need net poll to start new cycle */
+	if (!cycle->sc_cycle_started) {
+		D_ASSERT(cycle->sc_ults_tot == 0);
+		return true;
+	}
+
+	/* Need a nvme poll to end current cycle */
+	if (cycle->sc_ults_tot == 0)
+		return false;
+
+	/*
+	 * Need extra net poll when too many ULTs are processed in
+	 * current cycle.
+	 */
+	if (cycle->sc_age_net > cycle->sc_age_net_bound[1])
+		return true;
+
+	/* TODO: Take network request statistics into account */
+
+	return false;
+}
+
+static ABT_unit
+sched_pop_net_poll(struct sched_data *data, ABT_pool pool)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	ABT_unit		 unit;
+	int			 ret;
+
+	if (!need_net_poll(cycle))
+		return ABT_UNIT_NULL;
+
+	cycle->sc_age_net = 0;
+	cycle->sc_age_nvme++;
+	if (cycle->sc_ults_tot == 0) {
+		D_ASSERT(!cycle->sc_cycle_started);
+		cycle->sc_new_cycle = 1;
+	}
+
+	/* XXX Enabled this check once dss_srv_handler() is revised */
```

Comment:
for the possible revise of dss_srv_handler(), now the check_sleep_list() executed in it seems not very reasonable as the sleep ULT possibly belong to different pool. Maybe move it in the sched_run() loop or somewhere else?

Reply:
Yes, probably, but I guess Di didn't put it in sched_run() for some reason.